test/lit/passes/heap2local.wast

```diff
@@ -1945,4 +1980,56 @@
       )
     )
   )
+
+  ;; CHECK:      (func $cast-success (type $1) (result anyref)
+  ;; CHECK-NEXT:  (local $0 (ref $A))
+  ;; CHECK-NEXT:  (local $1 (ref $A))
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block (result nullref)
+  ;; CHECK-NEXT:    (local.set $1
+  ;; CHECK-NEXT:     (struct.new $A
+  ;; CHECK-NEXT:      (ref.null none)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (local.set $0
+  ;; CHECK-NEXT:     (local.get $1)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (ref.null none)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (local.get $0)
+  ;; CHECK-NEXT: )
+  (func $cast-success (result anyref)
+    (struct.get $A 0
+      (ref.cast (ref $A)
```

Comment:
This cast will be removed by other optimization passes (the cast can simply be removed), so while it's a good testcase I'm still uncertain I understand the situation in which this PR helps. The example from the PR description would also be handled by other opts (the local type would be refined).

Is there a case other optimizations can't help with?

Reply:
(To be clear, I do think it's very important we optimize these cases! It just seems like we might already do so.)

Reply:
You're right. This PR just makes the order in which passes are performed less relevant.
Currently, to get this optimization, since `heap2local` is performed before the casts are removed, i can repeat the `-O2` or `-O3` option (which can be slow on large programs) or perform the following passes at the end: `--heap2local --simplify-locals --reorder-locals --vacuum`.

Reply:
I see, thanks. In that case, perhaps we can add another run of `heap2local` in the main pipeline, after the casts are removed. We should measure that this isn't a significant regression in compile time, but otherwise we usually prefer that as a solution to such issues rather than add code complexity.

Reply:
I understand. I think one needs this sequence of passes: `local-subtyping`, `optimize-instructions`, `heap2local`,  `simplify-locals`, `reorder-locals`,  `vacuum`. So, several passes need to be added, but they look cheap.

Note that this PR does not much more than duplicating the existing code that handles `ref.as_non_null`.
https://github.com/WebAssembly/binaryen/blob/8c7ae2d9413fc6880da195b483c5a4e111c191fa/src/passes/Heap2Local.cpp#L649-L650

Reply:
No problem! Thanks! I guess I should have been clearer about that from the start...